core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractStatement.java

```diff
@@ -30,15 +35,15 @@ public boolean equals(Object o) {
 		// so these are checked last.
 
 		return this == o || o instanceof Statement
-				&& Objects.equals(getObject(), ((Statement) o).getObject())
-				&& Objects.equals(getSubject(), ((Statement) o).getSubject())
-				&& Objects.equals(getPredicate(), ((Statement) o).getPredicate())
+				&& getObject().equals(((Statement) o).getObject())
+				&& getSubject().equals(((Statement) o).getSubject())
+				&& getPredicate().equals(((Statement) o).getPredicate())
 				&& Objects.equals(getContext(), ((Statement) o).getContext());
 	}
 
 	@Override
 	public int hashCode() {
-		return Objects.hash(getSubject(), getPredicate(), getObject(), getContext());
+		return Objects.hash(getSubject(), getPredicate(), getObject(), getContext()); // !!! avoid array creation
```

Comment:
I genuinely don't understand what this comment means (also not sure I understand what the `!!!` signifies).

Reply:
`!!!` is just a leftover personal annotation, now removed

Reply:
`Objects.hash()` is a variadic method: invoking it requires the instantiation of an array containing arguments, which in the context of `.hasCode()` has a large performance impact (more than 25% according to my rough figures).